src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs

```diff
@@ -131,5 +138,103 @@ private static string GetHostnameFromStateObject(object hostNameOrAddress)
 
             return host;
         }
+
+        [NonEvent]
+        internal static string GetErrorType(Exception exception) => (exception as SocketException)?.SocketErrorCode switch
+        {
+            SocketError.HostNotFound => "host_not_found",
+            SocketError.TryAgain => "try_again",
+            SocketError.AddressFamilyNotSupported => "address_family_not_supported",
+            SocketError.NoRecovery => "no_recovery",
+
+            _ => exception.GetType().FullName!
+        };
+    }
+
+    /// <summary>
+    /// Encapsulates the starting timestamp together with an optional Activity, to represent the name resolution span for various telemetry pillars.
+    /// </summary>
+    internal readonly struct NameResolutionActivity
+    {
+        private const string ActivitySourceName = "Experimental.System.Net.NameResolution";
+        private const string ActivityName = ActivitySourceName + ".DnsLookup";
+        private static readonly ActivitySource s_activitySource = new(ActivitySourceName);
```

Comment:
NIT: my pet peeve, omitting the class with new makes looking for ctor calls impossible (as "find all references" is borked due to ref sources).
```suggestion
        private static readonly ActivitySource s_activitySource = new ActivitySource(ActivitySourceName);
```